### PR TITLE
fix(shared, nextjs): Support importing @clerk/shared in @clerk/backend

### PR DESCRIPTION
**Pull request by dimkl**
## Description

**Review it per commit**

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [x] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [x] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [x] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

**Review by anagstef — APPROVED** [review]
💯

File: packages/backend/src/redirections.test.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export default (QUnit: QUnit) => {
       assert.raises(
         () => redirectToSignIn({ returnBackUrl }),
         new Error(
-          'Missing publishableKey. You can get your key at https://dashboard.clerk.com/last-active?path=api-keys.',
+          '@clerk/backend: Missing publishableKey. You can get your key at https://dashboard.clerk.com/last-active?path=api-keys.',
```
[the remark this answers is not in the document]
Comment (anagstef) [reply packages/backend/src/redirections.test.ts:78]: Nvm, just saw this is a test file. My bad.

File: packages/shared/src/utils/callWithRetry.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+function wait(ms: number) {
+  return new Promise(res => setTimeout(res, ms));
+}
+
+const MAX_NUMBER_OF_RETRIES = 5;
+
+// TODO: Move this to @clerk/shared and reuse it with @clerk/clerk-js
```
[the remark this answers is not in the document]
Comment (dimkl) [reply packages/shared/src/utils/callWithRetry.ts]: yes it is 😅 . nice catch! i removed this.

File: packages/backend/src/util/shared.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  getClerkJsMajorVersionOrTag,
+  getScriptUrl,
+  isDevelopmentFromApiKey,
+  isProductionFromApiKey,
+  parsePublishableKey,
+} from '@clerk/shared';
+
+import { buildErrorThrower } from '@clerk/shared';
+// TODO: replace packageName with `${PACKAGE_NAME}@${PACKAGE_VERSION}` from tsup.config.ts
+export const errorThrower = buildErrorThrower({ packageName: '@clerk/backend' });
```
Comment (dimkl) [inline packages/backend/src/util/shared.ts:13]: I will fix this in another PR that will update how we build and run tests in backend package (currently tsup is not used in tests)

File: packages/shared/src/utils/keys.ts [hunk 4]
```diff
@@ -55,6 +55,7 @@ export function isLegacyFrontendApiKey(key: string) {
 }
 
 export function createDevOrStagingUrlCache() {
+  // TODO: Check if we can merge it with `./instance.ts#isStaging()`
   const DEV_OR_STAGING_SUFFIXES = [
```
Comment (dimkl) [inline packages/shared/src/utils/keys.ts:59]: I will cater this in another iteration where we will evaluate the exports of the `@clerk/shared`

File: packages/shared/src/utils/url.ts [hunk 5]
```diff
@@ -25,3 +27,28 @@ export function addClerkPrefix(str: string | undefined) {
   const stripped = str.replace(regex, '');
   return `clerk.${stripped}`;
 }
+
+export const getClerkJsMajorVersionOrTag = (frontendApi: string, pkgVersion?: string) => {
```
Comment (BRKalow) [inline packages/shared/src/utils/url.ts:31]: Can we add a brief JSDoc comment above this method as it's exported?

File: packages/shared/src/utils/url.ts [hunk 6]
```diff
@@ -25,3 +27,28 @@ export function addClerkPrefix(str: string | undefined) {
...
+  }
+
+  if (pkgVersion.includes('next')) {
+    return 'next';
+  }
+
+  return pkgVersion.split('.')[0] || 'latest';
+};
+
+export const getScriptUrl = (
```
Comment (BRKalow) [inline packages/shared/src/utils/url.ts:47]: Can we add a brief JSDoc comment above this method as it's exported?

File: packages/shared/src/utils/callWithRetry.ts [hunk 7]
```diff
@@ -4,7 +4,6 @@ function wait(ms: number) {
 
 const MAX_NUMBER_OF_RETRIES = 5;
 
-// TODO: Move this to @clerk/shared and reuse it with @clerk/clerk-js
 export async function callWithRetry<T>(
```
Comment (BRKalow) [inline packages/shared/src/utils/callWithRetry.ts:7]: Can we add a brief JSDoc comment above this method as it's exported?

**Review by BRKalow — COMMENTED** [review]
Can we add JSDoc comments to the exported methods moved to `@clerk/shared` as part of this? It should be quick, and leaves it better for the next person to touch this code 🙏

**Comment by dimkl** [comment]
@BRKalow JSDocs added to the moved helpers. ptal

**Comment by dimkl** [comment]
Since `@clerk/nextjs` depends on the `@clerk/backend` we should first merge this PR https://github.com/clerkinc/javascript/pull/1767 and then proceed with this. Otherwise we get the following error when building nextjs:
Screenshot 2023-09-27 at 17 22 49

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request merged**